wip: Network specification

mfornet opened the pull request:
Follow up from #12

SkidanovAlex commented on specs/ChainSpec/Messages.md:
```diff
@@ -0,0 +1,98 @@
+    /// decreased by 1. If this number is 0, drop this message.
+    ttl: u8,
+    /// Message
+    body: RoutedMessageBody,
```
I would add the definition of `RountedMessageBody` also to this file.

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+# Network
+
+Network layer constitutes the lower level of the NEAR protocol and is ultimately responsible of transporting messages between peers. To provide an efficient routing it maintains a routing table between all peers actively connected to the network, and send message between them using best paths. There is a mechanism in place that allows new peers joining the network to discover other peers, and reb ...
```
"and send message" -> "and sends messages"

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+## Discovering the network
+
+When a node start for the first time it tries to connect to a list of bootstrap nodes specified via a config file.
```
node start -> node starts

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+When a node start for the first time it tries to connect to a list of bootstrap nodes specified via a config file.
+
+It is expected that a node on a regular basis request a list of peers from its neighboring nodes to learn about other nodes in the network. This will allow every node to discover each other, and have relevant information to try to establish a new connection with it. When a node receive a message of type [`PeersRequest`](Messages.md#PeersMessage) it is expected to answer with a message of type [`P ...
```
request a list -> requests a list
when a node receive -> when a node receives

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+### Handshakes
+
+To establish new connections between pair of nodes, they will follow the following protocol. Node A open a connection with node B and sends a [Handshake](Messages.md#Handshake) to it. If handshake is valid (see reasons to [decline the handshake](#Decline-handshake)) then node B will proceed to send [Handshake](Messages.md#Handshake) to node A. After each node accept a handshake it will mark the o ...
```
Either "to establish new connections between pairs of nodes" or "to establish a new connection between a pair of nodes"

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+#### Edge
+
+Edges are used to let other know in the network that there is currently an active connection between a pair of nodes. See the signature of [this data structure](Messages.md#Edge).
```
"to let other know in the network" -> "to let others in the network know"
"the signature of this datastructure" -> "the definition of this datastructure" (to avoid confusion with cryptographic signatures)

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+Edges are used to let other know in the network that there is currently an active connection between a pair of nodes. See the signature of [this data structure](Messages.md#Edge).
+
+If the nonce of the edge is odd, it denotes an `Added` edge, otherwise it denotes a `Removed` edge. Each node should keep track of the nonce used for edges between every pair of nodes. Peer C believe that peer A and B are currently connected if and only if the edge with higher nonce known to C about them have odd nonce.
```
"Peer C believe that peer A and B are currently connected if and only if the edge with higher nonce known to C about them have odd nonce." -> "Peer C believes that the peers A and B are currently connected if and only if the edge with the highest nonce known to C for them has an odd nonce."

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+When two nodes successfully connect to each other, they broadcast the new edge to let other peers know about this connection. When a node is disconnected from other node, it should bump the nonce by 1, sign the new edge and broadcast it to let other nodes know that the connection was removed.
+
+A removed connection will be valid, if it contains valid information from the added edge it is invalidating. This will prevent peers bump nonce by more than one when deleting an edge.
```
This will prevent -> This prevents

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+A removed connection will be valid, if it contains valid information from the added edge it is invalidating. This will prevent peers bump nonce by more than one when deleting an edge.
+
+When node A propose an edge to B with nonce X, it will only accept it and sign it if:
```
propose -> proposes
also, if -> iff?

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+When node A propose an edge to B with nonce X, it will only accept it and sign it if:
+
+- X = 1 and B doesn't know about any previous edge between A and B
+- X is odd and X > Y where Y is the nonce of the edge with higher nonce between A and B known to B.
```
higher -> the highest

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+## Routing Table
+
+Every node will maintain a routing table with all existing connections and relevant information to route messages. The explicit graph with all active connection is stored at all times.
```
will maintain -> maintains

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
````diff
@@ -0,0 +1,119 @@
+}
+```
+
+`RoutingTable` should be update accordingly the node receives updates from the network. There are four types of updates `RoutingTable` cares about:
````
accordingly the -> accordingly when the

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+### Routing
+
+When a node needs to send a message to another peer, it checks in the routing table if it is connected to that peer, possibly not directed but through several hops. Then it select one of the shortest path to the target peer and send a [`RoutedMessage`](Messages.md#RoutedMessage) to the first peer in the path.
```
possibly not directed -> possibly not directly

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+When a node needs to send a message to another peer, it checks in the routing table if it is connected to that peer, possibly not directed but through several hops. Then it select one of the shortest path to the target peer and send a [`RoutedMessage`](Messages.md#RoutedMessage) to the first peer in the path.
+
+When it receives a `RoutedMessage`, it check if it is the target, in that case consume the body of the message, otherwise it finds a route to the target following described approach and send the message again. Is is important that before routing a message each peer check signature from original author of the message, passing a message with invalid signature can result in ban for sender. It is not ...
```
and send the message again -> and sends the message again
in ban for sender -> in ban for the sender

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+When it receives a `RoutedMessage`, it check if it is the target, in that case consume the body of the message, otherwise it finds a route to the target following described approach and send the message again. Is is important that before routing a message each peer check signature from original author of the message, passing a message with invalid signature can result in ban for sender. It is not ...
+
+Each `RoutedMessage` is equipped with a time-to-live integer. If this message is not for us, we decrement by one this field, if value is 0, we drop this message, otherwise we forward it.
```
"If this message is not for the node processing it, it decrement the field by one before routing it; if the value is 0, the node drops the message instead of routing it."

Generally, avoid using "we" and "us".

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+#### Routing back
+
+It is possible that node A is known to B but not the other way around. In case node A sends a request that requires a response to B, the response is routed back through the same path used to send the message from A to B. This are special types of messages.
```
> This are special types of messages

Those are not really messages, right, those are some sort of routines?

mfornet replied:
Reworded

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+## Security
+
+Messages exchanged between routed (both direct or routed) are cryptographically signed with sender private key. Nodes ID contains public key of each node that allow other peer to verify this messages. To keep secure communication most of this message requires some nonce/timestamp that forbid a malicious actor reuse a signed message out of context.
```
between routed -> between peers?

SkidanovAlex commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,119 @@
+
+### Abusive behavior
+
+When a node A sends more than `MAX_PEER_MSG_PER_MIN` messages per minute to node B, it will be banned and unable to keep sending messages to it. This a protection mechanism against abusive node to avoid being spammed by some peers.
```
This should be phrased as a recommendation, since it is not a requirement for the node to be protocol compliant, and other implementations might choose to use different protection mechanisms.

mfornet replied:
Yes, but how it is implemented might affect other nodes in the network. Even if it is not mandatory, it is desired to know rules that will make my node being banned, so that I avoid violating them.

SkidanovAlex commented:
Lots of bikeshedding inline.

I think the level of detalization of the routing is too low. In particular, it doesn't quite explain what is the key in the routing table, how to send a message back via routing.

SkidanovAlex commented:
Also missing, but important:

- Definitions of `PeerId` and `PeerInfo`
- What port do we connect to? Does it matter, or the node can use any?

SkidanovAlex commented:
Also the spec appears to miss a critical piece of information that the messages are not just Borsh, they are 4 bytes length followed by the borshified message.

mfornet commented:
> * What port do we connect to? Does it matter, or the node can use any?

They can use any. They will learn about the ip/port used by other nodes through the PeerDiscovery mechanism.

SkidanovAlex commented:
> They can use any. They will learn about the ip/port used by other nodes through the PeerDiscovery mechanism.

What I mean is that it should be clearer in the spec :)

mfornet commented:
> Also the spec appears to miss a critical piece of information that the messages are not just Borsh, they are 4 bytes length followed by the borshified message.

Isn't that part of borsh standard?

- Decoding messages: https://github.com/nearprotocol/nearcore/blob/master/chain/network/src/peer.rs#L585

- Encoding message: https://github.com/nearprotocol/nearcore/blob/master/chain/network/src/peer.rs#L220

We let borsh library to encode/decode messages for us, without handling any 4 bytes length prefix.

SkidanovAlex commented:
> Isn't that part of borsh standard?

The way I see it, `send_message` pipes the output of `peer_message_to_bytes` into `framed`, which in turn uses `Codec`, which in turn prepends the four bytes. None of that appears to be part of borsh spec.

Though I actually don't see where we use the decoder from the `Codec`

mfornet commented:
> > Isn't that part of borsh standard?
> 
> The way I see it, `send_message` pipes the output of `peer_message_to_bytes` into `framed`, which in turn uses `Codec`, which in turn prepends the four bytes. None of that appears to be part of borsh spec.
> 
> Though I actually don't see where we use the decoder from the `Codec`

I see. We are using the decoder here: https://github.com/nearprotocol/nearcore/blob/master/chain/network/src/peer_manager.rs#L358

mfornet commented:
@SkidanovAlex just addressed all your comments, and added some pseudo-code. I plan to add a diagram for the handshake process, since it is kind of technical/complex and under-specified.

bowenwang1996 commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,211 @@
+
+A removed connection will be valid, if it contains valid information from the added edge it is invalidating. This prevents peers bump nonce by more than one when deleting an edge.
+
+When node A proposes an edge to B with nonce X, it will only accept it and sign it iff:
```
```suggestion
When node A proposes an edge to B with nonce X, B will only accept it and sign it iff:
```

bowenwang1996 commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,211 @@
+
+When a node A sends more than `MAX_PEER_MSG_PER_MIN` messages per minute to node B, it will be banned and unable to keep sending messages to it. This a protection mechanism against abusive node to avoid being spammed by some peers.
+
+## Possible attacks
```
Let's not discuss attacks here

bowenwang1996 commented on specs/ChainSpec/NetworkSpec.md:
```diff
@@ -0,0 +1,211 @@
+
+### Balancing network
+
+[Github comment](https://github.com/nearprotocol/nearcore/issues/2395#issuecomment-610077017)
```
We should not just post a link here if it is supposed to be part of the spec
